blockchain/v2/reactor.go

```diff
@@ -129,15 +134,16 @@ type blockStore interface {
 type BlockchainReactor struct {
 	p2p.BaseReactor
 
-	fastSync  bool       // if true, enable fast sync on start
-	events    chan Event // XXX: Rename eventsFromPeers
-	scheduler *Routine
-	processor *Routine
-	logger    log.Logger
+	fastSync    bool // if true, enable fast sync on start
+	stateSynced bool // set to true when SwitchToFastSync is called by state sync
+	scheduler   *Routine
+	processor   *Routine
+	logger      log.Logger
 
 	mtx           sync.RWMutex
 	maxPeerHeight int64
 	syncHeight    int64
+	events        chan Event // non-nil during a fast sync
```

Comment:
why make it nil? can't we use another variable under mtx to signal sync in progress. then we won't need to lock/unlock mtx every time we access events

Reply:
If we have a different variable we'll need a mutex around that, since the variable will be updated from separate goroutines.  We also close the channel to terminate the event loop, making the channel unusable anyway. We could use an atomic, but that introduces a race condition where the variable is true, then there's a context switch closing the channel, then the original routine tries to write to the closed channel and panics. I feel like it's safer to just completely remove it after it's been closed.

Reply:
do we have to close the channel?

Reply:
Closing channels is a pretty standard way to signal worker/eventloop termination. If we leave the channel open but there are no receivers, then messages may accumulate in it - once full, it will block senders, which will block p2p goroutines, which will cause peers to be disconnected.